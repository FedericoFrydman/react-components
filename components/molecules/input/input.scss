@import '../../assets/styles/fonts.scss';
@import '../../assets/styles/colors.scss';

@mixin createBorder($color) {
  border: 1px solid $color;
  -webkit-border-color: none;
  border-radius: var(--brand-border-radius);
  -moz-box-sizing: border-box; /* Firefox */
  -webkit-box-sizing: border-box; /* Safari/Chrome */
  box-sizing: border-box;
}

@mixin focusState($color) {
  border-color: $color;
}

.lsux-input-container {
  flex: 1;
  position: relative;

  &__input {
    background-color: $N00;
    -moz-box-sizing: border-box; /* Firefox */
    -webkit-box-sizing: border-box; /* Safari/Chrome */
    box-sizing: border-box;
    color: $N800;
    @include createBorder($N350);
    font-style: normal;
    height: 44px;
    letter-spacing: -0.011em;
    line-height: 22px;
    outline: none;
    padding-left: 16px;
    padding-right: 16px;
    -moz-transition: border-color ease-in-out 0.15s; /* Firefox */
    -webkit-transition: border-color ease-in-out 0.15s; /* Safari/Chrome */
    transition: border-color ease-in-out 0.15s;
    width: 100%;

    &:focus:not(:read-only) {
      @include focusState(var(--brand-color-prim200));
      outline: none;
    }

    &--warning {
      background: $Y10;
      outline: none;
      @include createBorder($Y200);

      &:focus {
        @include focusState($Y200);
      }
    }

    &--valid {
      @include createBorder($G200);
      outline: none;

      &:focus {
        @include focusState($G200);
      }
    }

    &--invalid {
      background: $R10;
      outline: none;
      @include createBorder($R200);

      &:focus {
        @include focusState($R200);
      }

      &:not(:focus) {
        color: $R200;
        & + .lsux-text {
          color: $R200;
        }
      }
    }

    &--icon-left {
      padding-left: 40px;
    }

    &--icon-right {
      padding-right: 40px;
    }

    &:disabled {
      background-color: $N50;
      color: $N350;
      cursor: not-allowed;
    }

    &:-moz-read-only {
      border: none;
      color: $N700;
    }

    &:read-only {
      border: none;
      color: $N700;
    }

    &::-webkit-inner-spin-button {
      background: $N100;
      border: 1px solid $N500;
      border-radius: var(--brand-border-radius);
      height: 40px;
      margin: 0;
      opacity: 1;
    }

    &::-webkit-calendar-picker-indicator {
      background: none;
      cursor: pointer;
      display: block;
      height: 20px;
      margin-right: 10px;
      outline: none;
      width: 20px;
    }

    &::placeholder {
      align-items: center;
      color: $N500;
      display: flex;
      right: 8px;
      top: 8px;
    }

    &--date-number::-webkit-inner-spin-button,
    &--date-number::-webkit-outer-spin-button {
      -webkit-appearance: none;
      border: none;
      margin: 0;
    }
  }

  .lsux-input__icon {
    top: 8px;
    &-left {
      left: 8px;
    }
    &-right {
      right: 8px;
    }
    &-hide-icon {
      border-radius: var(--brand-border-radius);
      bottom: 0;
      position: absolute;
      right: 0;
      top: 0;
      width: 34px;
      img {
        top: 12px;
      }
    }
    &-search-icon {
      left: 8px;
      position: absolute;
      top: 8px;
    }
  }

  .lsux-button {
    background-color: transparent;
  }

  .lsux-icon {
    cursor: pointer;
    position: absolute;
    right: 8px;
    top: 10px;
  }
}
