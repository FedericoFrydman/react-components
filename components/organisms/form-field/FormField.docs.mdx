import { FC } from 'react';
import { ArgsTable, Story, Canvas } from '@storybook/addon-docs';
import { Alert } from '../../molecules/alert/Alert';

# Form Field

The FormField component is a container that adds the labels; validation messages, hints and colors for the different controls in a form (inputs, selects, checkbox, radio).

<Alert title="Dev Note" severity="warning" icon="alert_warning">
  Based on the text above, Input component should always be used as child of FormField.
</Alert>

<p style={{ margin: 0 }}>&nbsp;</p>

<ArgsTable />

**About the input:**

- Input allows people to enter data for further use in your product, or to configure options. Inputs and forms are used to get information and guide people with minimal fuss.
- Input should always be accompanied by a label for accessibility purposes with the exception of Search (for this reason, the default text color for Search is N1000 for color contrast.

## Type

<p style={{ margin: 0 }}>&nbsp;</p>

### Text Field

<Canvas>
  <Story id="controls-form-fields--text-field" />
</Canvas>

### Number

<Canvas>
  <Story id="controls-form-fields--number-field" />
</Canvas>

- Up and down buttons to make it easy for user to adjust amount
- Can use “up” or “down” keyboard to adjust amount when input is focused

### Password

<Canvas>
  <Story id="controls-form-fields--password-field" />
</Canvas>

### Tel

<Canvas>
  <Story id="controls-form-fields--tel-field" />
</Canvas>

### TextArea

<Canvas>
  <Story id="controls-form-fields--text-area-field" />
</Canvas>

- Consider to add character counter or add description to show character limit
- Auto increase height or can also be adjusted by user

### Search

<Canvas>
  <Story id="controls-form-fields--search-field" />
</Canvas>

### Date

<Canvas>
  <Story id="controls-form-fields--date" />
</Canvas>

- These inputs has the format for birth day, that means month/day/year. Click on Show code to see how to implement this.

<Alert title="Dev Note" severity="warning" icon="alert_warning">
  In case you need just one date input, have in mind you can set the input type as a date and its behavior would be the
  default browser behavior.
</Alert>

### Select

<Canvas>
  <Story id="controls-form-fields--select-field" />
</Canvas>

### Radio

<Canvas>
  <Story id="controls-form-fields--radio-field" />
</Canvas>

### Checkbox

<Canvas>
  <Story id="controls-form-fields--checkbox-field" />
</Canvas>

### Required

<Canvas>
  <Story id="controls-form-fields--input-required" />
</Canvas>

### Description

<Canvas>
  <Story id="controls-form-fields--description" />
</Canvas>

## Usage guidelines

- **Do** include a label with all inputs. Label should be sentence case and keep it short
- **Do not** use placeholder as label text
- **Do** use placeholder text to give hints what should go into the input/field
- **Do** make it clear with an \* asterisk if an individual input is required
- **Do** mark "Optional" for optional inputs when majority of other inputs are required
- **Do** include description if an input format is expected or give context why we are collecting the information from users

## States

<p style={{ margin: 0 }}>&nbsp;</p>

In order to display the input state you must define the status prop and use the validationHint prop as well for user experience.

### Invalid/Error

<Canvas>
  <Story id="controls-form-fields--invalid" />
</Canvas>

### Warning

<Canvas>
  <Story id="controls-form-fields--warning" />
</Canvas>

### Success

<Canvas>
  <Story id="controls-form-fields--success" />
</Canvas>

### Disabled

<Canvas>
  <Story id="controls-form-fields--disabled" />
</Canvas>

### Read only

<Canvas>
  <Story id="controls-form-fields--read-only" />
</Canvas>

## Example Code

```tsx
export const EmailForm: FC = ({}) => {

  ...

  return (
    <Form onSubmit={submitForm}>
      <FormField
        id="email"
        status={email.status}
        validationHint={email.hint}
        label="Email Address"
        type="input"
        required
      >
        <Input
          name="email"
          onChange={handleInputChange}
          placeholder="Email Address"
          value={email.value}
          pattern={EMAIL_PATTERN}
          type="email"
          required
        />
      </FormField>
      <Button
        type="submit"
        disabled={!isValid}
        label="Send Email"
        classNames={['mt-5']}
        stretch
      />
    </Form>
  )
}
```
