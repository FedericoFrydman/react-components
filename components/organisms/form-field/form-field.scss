@import '../../assets/styles/fonts.scss';
@import '../../assets/styles/colors.scss';

.lsux-form-field-container {
  margin-top: 24px;
  position: relative;
  width: auto;

  &--stretch {
    width: 100%;
  }

  &--no-margin {
    margin: 0;
  }

  &--day,
  &--month {
    width: 60px;
  }

  &--year {
    width: 112px;
  }

  &__label {
    color: $N700;
    display: block;
    font: $T14;
    font-weight: normal;
    left: 0;
    letter-spacing: -0.006em;
    line-height: 18px;
    margin-bottom: 8px;
    top: 0;

    &.disabled {
      color: $N500;
    }

    .required {
      color: $R100;
      line-height: 0;
      margin-left: 3px;
    }
  }

  &__description {
    color: $N700;
    display: block;
    margin-top: 8px;
  }

  &__hint {
    align-items: center;
    display: flex;
    flex-direction: row;

    .lsux-icon {
      height: 14px;
    }

    &__text {
      color: $N700;

      &--invalid {
        color: $R200;
      }
    }
  }
}
