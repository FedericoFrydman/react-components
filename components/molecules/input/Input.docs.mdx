import { ArgsTable, Story, Canvas, Stories } from '@storybook/addon-docs';
import { Alert } from '../alert/Alert';
import { Input } from './Input';

# Input Props

**This documentantion is just to display the input props where you can interact with them and see how it behaves. For implementation please use [FormField](/?path=/docs/forms-form-fields--input-field) docs to gide.**

<Alert title="Dev Note" severity="warning" icon="alert_warning">
  The Input component must be used as a child of a FormField component.
</Alert>

<p style={{ margin: 0 }}>&nbsp;</p>

<ArgsTable of={Input} />

**About the input:**

- Input allows people to enter data for further use in your product, or to configure options. Inputs and forms are used to get information and guide people with minimal fuss.
- Input should always be accompanied by a label for accessibility purposes with the exception of Search (for this reason, the default text color for Search is N1000 for color contrast.
- Input has a HTML property call autoComplete it should be equal to "off" in order to not be auto completed.

## Example Code

```tsx
<FormField id="name-input" status={name.status} validationHint={name.hint} label="First Name" type="input" required>
  <Input
    name="name-input"
    onChange={handleInputChange}
    placeholder="First Name"
    value={name.value}
    type="text"
    required
  />
</FormField>
```

<Stories />
