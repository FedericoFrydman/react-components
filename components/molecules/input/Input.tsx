import { cloneElement, DetailedHTMLProps, forwardRef, ReactElement, Ref, useEffect, useState } from 'react';
import NumberFormat, { NumberFormatProps } from 'react-number-format';
import cn from 'classnames';

import { Button } from '../button/Button';
import { Icon, IconNames, IconProps } from '../../atoms/icon/Icon';
import { IconColors } from '../../../components/assets/icons';
import './input.scss';

export type InputValidationStatus = '' | 'valid' | 'warning' | 'invalid';

export type InputType =
  | 'color'
  | 'date'
  | 'datetime-local'
  | 'email'
  | 'hidden'
  | 'month'
  | 'number'
  | 'password'
  | 'range'
  | 'search'
  | 'tel'
  | 'text'
  | 'time'
  | 'url'
  | 'week'
  | 'currency';

export interface InputProps extends Omit<React.HTMLProps<HTMLInputElement>, 'type'> {
  /**
   * Triggers error | warning | success styles on the component. Important for accessibility.
   */
  status?: InputValidationStatus;
  /**
   * Determines the type of the Input.
   * @default text
   */
  type?: InputType;
  /**
   * Determines which icon should be shown on the right; if a password type this property is managed internally with the hide/show icons.
   * @deprecated Please use `iconLeft` and `iconRight` instead.
   */
  icon?: IconNames;
  /**
   * Icon on the left side of the input
   */
  iconLeft?: ReactElement<IconProps>;
  /**
   * Icon on the right side of the input
   */
  iconRight?: ReactElement<IconProps>;
  /**
   * Applies a color to the icon.  This could be for the icon on the right or a search icon when type is search.
   * @default BRAND200
   */
  iconColor?: IconColors;
  /**
   * Custom class names for the Input.
   */
  classNames?: cn.Argument;
  /**
   * Format uses the NumberFormat component instead of Input. It allows for input masking.
   */
  format?: string;
  /**
   * Placeholder value for the input field.
   */
  placeholder?: string;
  /**
   * Prefix value for the input field. Only displays when `NumberFormat` component is rendered.
   */
  prefix?: string;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLInputElement>;
}

const InputComponent = (
  {
    status = '',
    type = 'text',
    icon,
    iconColor = 'BRAND200',
    format,
    placeholder,
    prefix,
    iconLeft,
    iconRight = icon ? <Icon name={icon} color={iconColor} /> : null,
    classNames = [],
    ...props
  }: InputProps,
  ref: InputProps['ref'],
) => {
  const [iconColorState, setIconColorState] = useState(iconColor);
  const [passwordVisible, setPasswordVisible] = useState(false);

  const classInputContainer = `lsux-input-container`;
  const cnInputContainer = cn(classInputContainer);

  const classInput = 'lsux-input-container__input';
  const cnInput = cn(
    {
      [`${classInput}`]: true,
      [`${classInput}--${status}`]: status,
      [`${classInput}--icon-left`]: type === 'search' || iconLeft,
      [`${classInput}--icon-right`]: type === 'password' || iconRight,
    },
    classNames,
  );

  const classInputIcon = 'lsux-input__icon';

  useEffect(() => {
    switch (status) {
      case 'invalid': {
        setIconColorState('R200');
        break;
      }
      case 'warning': {
        setIconColorState('Y200');
        break;
      }
      case 'valid': {
        setIconColorState('G200');
        break;
      }
      default: {
        setIconColorState(iconColor);
        break;
      }
    }
  }, [status, iconColor]);

  if (type === 'password' && !iconColor) {
    setIconColorState('BRAND200');
  }

  if (iconLeft) {
    iconLeft = cloneElement(iconLeft, {
      classNames: [iconLeft.props?.classNames, classInputIcon, `${classInputIcon}-left`],
      color: iconColorState,
    });
  }

  if (iconRight) {
    iconRight = cloneElement(iconRight, {
      classNames: [iconRight.props?.classNames, classInputIcon, `${classInputIcon}-right`],
      color: iconColorState,
    });
  }

  if (type === 'search') {
    iconLeft = (
      <Icon
        name="interface_search_magnifying_glass"
        classNames={[classInputIcon, `${classInputIcon}-search-icon`]}
        color={iconColorState}
      />
    );
  }

  if (type != 'password' && icon) {
    iconRight = <Icon name={icon} color={iconColorState} classNames={classInputIcon} />;
  }

  const handleIconClick = () => {
    setPasswordVisible(!passwordVisible);
  };

  const { ref: removedRef, ...other } = props;

  // Set default format value for `tel` input
  if (type === 'tel' && !format) {
    format = '(###) ###-####';
  }

  // Set default placeholder value for `tel` input
  if (type === 'tel' && !placeholder) {
    placeholder = '(___) ___-____';
  }

  const content = (
    <div className={cnInputContainer}>
      {iconLeft}
      {format && (type === 'password' || type === 'tel' || type === 'text') && (
        <NumberFormat
          className={cnInput}
          format={format}
          mask="_"
          placeholder={placeholder}
          prefix={prefix}
          type={type}
          {...(props as NumberFormatProps)}
        />
      )}
      {type === 'currency' && (
        <NumberFormat
          allowLeadingZeros={false}
          className={cnInput}
          format={format}
          placeholder={placeholder ?? `${prefix ?? '$'}0.00`}
          prefix={prefix ?? '$'}
          thousandSeparator={true}
          type="text"
          {...(props as NumberFormatProps)}
        />
      )}
      {!format && type !== 'currency' && (
        <input
          {...(props as DetailedHTMLProps<unknown, unknown>)}
          type={type === 'password' && passwordVisible ? 'text' : type}
          placeholder={placeholder}
          className={cnInput}
          ref={ref}
        />
      )}
      {type === 'password' && (
        <Button
          classNames={[classInputIcon, `${classInputIcon}-hide-icon`]}
          iconColor={iconColorState}
          iconLeft={passwordVisible ? 'action_eye_closed' : 'action_eye_open'}
          onClick={handleIconClick}
          variant="tertiary"
        />
      )}
      {type !== 'password' && iconRight}
    </div>
  );

  return content;
};

export const Input = forwardRef<HTMLInputElement, InputProps>(InputComponent);
