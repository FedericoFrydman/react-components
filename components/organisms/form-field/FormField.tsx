import { Children, forwardRef, ReactElement, Ref, useEffect, useState } from 'react';
import cn from 'classnames';

import './form-field.scss';
import { Text } from '../../atoms/text/Text';
import { Icon, IconNames } from '../../atoms/icon/Icon';
import { InputValidationStatus } from '../../molecules/input/Input';
import { IconColors } from '../../../components/assets/icons';

export interface FormFieldProps {
  /**
   * A unique identifier for the input field.
   */
  id?: string;
  /**
   * A clear and concise description of the form-field purpose.
   */
  label?: string;
  /**
   * A label to the left of the FormField
   */
  leftLabel?: string;
  /**
   * A label to the right of the FormField
   */
  rightLabel?: string;
  /**
   * Warning or error message.
   */
  validationHint?: string;
  /**
   * Triggers error | warning | success styles on the component. Important for accessibility.
   */
  status?: InputValidationStatus;
  /**
   * Brief description under the component
   */
  description?: string;
  /**
   * Removes the default margin below the text
   */
  noMargin?: boolean;
  /**
   * Children components
   */
  children?: ReactElement[] | ReactElement;
  /**
   * Stretch the component across the full width of its parent.
   */
  stretch?: boolean;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Define if the input is required
   */
  required?: boolean;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

/**
 * The FormField component is a container that adds the labels; validation messages, hints and colors for the different controls in a form (inputs, selects, checkbox, radio).
 */
const FormFieldComponent = (
  {
    id,
    label,
    leftLabel,
    rightLabel,
    validationHint,
    description,
    status,
    noMargin,
    children,
    stretch,
    classNames = [],
    required = false,
    ...props
  }: FormFieldProps,
  ref: FormFieldProps['ref'],
) => {
  const classFormField = 'lsux-form-field';
  const classContainer = `${classFormField}-container`;
  const cnContainer = cn(
    classContainer,
    {
      [`${classContainer}--stretch`]: stretch,
      [`${classContainer}--no-margin`]: noMargin,
    },
    classNames,
  );

  const cnValidation = {
    [`${classContainer}__hint_text`]: status == 'valid' || status == 'warning',
    [`${classContainer}__hint_text--invalid`]: status == 'invalid',
    ['pl-2']: true,
  };

  const [validationIcon, setValidationIcon] = useState<IconNames>(null);
  const [validationColor, setValidationColor] = useState<IconColors>(null);
  const [disabled, setDisabled] = useState(false);

  useEffect(() => {
    if (validationHint) {
      switch (status) {
        case 'valid':
          setValidationIcon('alert_success');
          setValidationColor('G200');
          break;
        case 'warning':
          setValidationIcon('alert_warning');
          setValidationColor('Y300');
          break;
        case 'invalid':
          setValidationIcon('alert_information');
          setValidationColor('R300');
          break;
        default:
          setValidationIcon(null);
          setValidationColor(null);
          break;
      }
    }
  }, [status, validationHint]);

  const elements = Children.toArray(children);

  useEffect(() => {
    if (elements.length > 0) {
      const element = elements[0] as ReactElement;
      setDisabled(element.props.disabled);
    }
  }, [elements]);

  return (
    <div {...props} ref={ref} className={cnContainer}>
      {label && (
        <label className={cn(`${classContainer}__label`, { disabled: disabled })} htmlFor={id}>
          {label}
          {required && <span className="required">*</span>}
        </label>
      )}
      {rightLabel || leftLabel ? (
        <div style={{ alignItems: 'center', display: 'flex' }}>
          {leftLabel && <div className="mr-3">{leftLabel}</div>}
          {children}
          {rightLabel && <div className="ml-3">{rightLabel}</div>}
        </div>
      ) : (
        <>{children}</>
      )}
      {description && (
        <Text textSize="small" classNames={[`${classContainer}__description`, 'pt-3']} text={description} />
      )}
      {validationHint && (
        <div className={cn(`${classContainer}__hint`, `pt-3`)}>
          <Icon name={validationIcon} color={validationColor} size="small" />
          <Text textSize="small" classNames={cnValidation} text={validationHint} />
        </div>
      )}
    </div>
  );
};

export const FormField = forwardRef(FormFieldComponent);
