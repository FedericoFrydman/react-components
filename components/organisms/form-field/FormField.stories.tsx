import { FC } from 'react';
import { Story, Meta } from '@storybook/react';

import { FormField, FormFieldProps } from './FormField';

import docs from './FormField.docs.mdx';
import { Input } from '../../molecules/input/Input';
import { Checkbox } from '../../molecules/checkbox/Checkbox';
import { Radio } from '../../molecules/radio/Radio';
import { Select, SelectItem } from '../../molecules/select/Select';
import { TextArea } from '../../atoms/text-area/TextArea';
import { Container } from '../../atoms/container/Container';
import { MultiSelect, SelectOption } from '../../molecules/multi-select/MultiSelect';

const options: SelectOption[] = [
  {
    label: 'United States',
    value: 'US',
  },
  {
    label: 'Germany',
    value: 'DE',
  },
  {
    label: 'France',
    value: 'FR',
  },
];

const FormFieldStory: Story<FormFieldProps> = (args: FormFieldProps) => (
  <Container classNames={['m-4']}>
    <FormField {...args} />
  </Container>
);

export const TextField = FormFieldStory.bind({});
TextField.args = {
  children: <Input placeholder="First Name" name="name" />,
  id: 'base-input',
  label: 'First Name',
};

export const NumberField = FormFieldStory.bind({});
NumberField.args = {
  children: <Input placeholder="0" name="number" type="number" />,
  id: 'number-input',
  label: 'Number',
};

export const TextAreaField = FormFieldStory.bind({});
TextAreaField.args = {
  children: <TextArea placeholder="Message" rows={5} />,
  id: 'base-text-area',
  label: 'Message',
};

export const SearchField = FormFieldStory.bind({});
SearchField.args = {
  children: <Input placeholder="Search" name="search-input" type="search" />,
  id: 'search-input',
  label: 'Search',
};

export const TelField = FormFieldStory.bind({});
TelField.args = {
  children: <Input type="tel" format="### ###-####" placeholder="123-456-789" />,
  id: 'input-mask',
  label: 'Tel',
};

export const PasswordField = FormFieldStory.bind({});
PasswordField.args = {
  children: <Input type="password" placeholder="********" />,
  id: 'input-password',
  label: 'Password',
};

export const CurrencyField = FormFieldStory.bind({});
CurrencyField.args = {
  children: <Input type="currency" />,
  id: 'input-currency',
  label: 'Currency',
};

export const SelectField = FormFieldStory.bind({});
SelectField.args = {
  children: (
    <Select placeholder="Select a country">
      <SelectItem textValue="United States" value="usa" />
      <SelectItem textValue="Canada" value="canada" />
      <SelectItem textValue="Mexico" value="mexico" />
    </Select>
  ),
  id: 'base-select',
  label: 'Country',
};

export const MultiSelectField = FormFieldStory.bind({});
MultiSelectField.args = {
  children: <MultiSelect placeholder="Select a country" options={options} />,
  id: 'base-multi-select',
  label: 'Country',
};

export const RadioField = FormFieldStory.bind({});
RadioField.args = {
  children: <Radio rightLabel="Radio Label" />,
  id: 'base-radio',
};

export const CheckboxField = FormFieldStory.bind({});
CheckboxField.args = {
  children: <Checkbox rightLabel="Checkbox Label" name="checkbox-field" readOnly />,
  id: 'base-checkox',
};

export const InputRequired = FormFieldStory.bind({});
InputRequired.args = {
  children: <Input type="text" placeholder="Required example" required />,
  id: 'input-required',
  label: 'Input required label',
  required: true,
};

export const Disabled = FormFieldStory.bind({});
Disabled.args = {
  children: <Input type="text" placeholder="Disabled input" disabled={true} />,
  id: 'input-disabled',
  label: 'Disabled input',
};

export const ReadOnly = FormFieldStory.bind({});
ReadOnly.args = {
  children: <Input type="text" value="Some read only value" readOnly={true} placeholder="Some read only value" />,
  id: 'input-read-only',
  label: 'Read only',
};

export const Invalid = FormFieldStory.bind({});
Invalid.args = {
  children: <Input placeholder="First Name" status="invalid" />,
  id: 'invalid-input',
  label: 'First Name',
  status: 'invalid',
  validationHint: 'This field has an error',
};

export const Warning = FormFieldStory.bind({});
Warning.args = {
  children: <Input placeholder="First Name" status="warning" />,
  id: 'warning-input',
  label: 'First Name',
  status: 'warning',
  validationHint: 'Warning message',
};

export const Success = FormFieldStory.bind({});
Success.args = {
  children: <Input placeholder="First Name" status="valid" />,
  id: 'success-input',
  label: 'First Name',
  status: 'valid',
  validationHint: 'Success message',
};

export const Description = FormFieldStory.bind({});
Description.args = {
  children: <Input placeholder="First Name" />,
  description: 'This field has a description',
  id: 'description-input',
  label: 'First Name',
};

export const Date: FC = () => (
  <Container classNames={['m-4']}>
    <FormField id="input--birth-date--month" label="Month">
      <Input placeholder="MM" type="number" />
    </FormField>
    <FormField id="input--birth-date--day" label="Day">
      <Input placeholder="DD" type="number" />
    </FormField>
    <FormField id="input--birth-date--year" label="Year">
      <Input placeholder="YYYY" type="number" />
    </FormField>
  </Container>
);

export default {
  argTypes: {
    description: { control: 'text' },
    label: { control: 'text' },
    leftLabel: { control: 'text' },
    noMargin: { control: 'boolean' },
    required: { control: 'boolean' },
    rightLabel: { control: 'text' },
    status: { control: 'select', options: ['', 'valid', 'warning', 'invalid'] },
    stretch: { control: 'boolean' },
    validationHint: { control: 'text' },
  },
  component: FormFieldStory,
  parameters: {
    docs: { page: docs },
  },
  subcomponents: {
    Checkbox,
    Input,
    Radio,
    Select,
  },
  title: 'Controls/Form Fields',
} as Meta;
