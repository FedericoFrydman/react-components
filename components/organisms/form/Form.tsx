import { Ref, HTMLProps, forwardRef, ReactElement } from 'react';
import cn from 'classnames';

import { FormFieldProps } from '../form-field/FormField';

export interface FormProps extends Omit<HTMLProps<HTMLFormElement>, 'css'> {
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Controls of the form
   */
  children: ReactElement<FormFieldProps> | ReactElement<FormFieldProps>[];
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLFormElement>;
}

const FormComponent = ({ classNames = [], children, ...props }: FormProps, ref: FormProps['ref']) => {
  return (
    <form {...props} className={cn('lsux-form', classNames)} ref={ref}>
      {children}
    </form>
  );
};

/**
 * The Form component works as a wrapper for FormFields. It has all the properties and events of a HTML form so it can handle the validity state of the form and the onSubmit event.
 */

export const Form = forwardRef<HTMLFormElement, FormProps>(FormComponent);
