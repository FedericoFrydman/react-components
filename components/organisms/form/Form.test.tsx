import { shallow, ShallowWrapper } from 'enzyme';
import { Form } from './Form';
import { Input } from '../../molecules/input/Input';
import { FormField } from '../../organisms/form-field/FormField';

describe('Form Component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Form>
        <FormField id="input" label="text">
          <Input />
        </FormField>
      </Form>,
    );
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
