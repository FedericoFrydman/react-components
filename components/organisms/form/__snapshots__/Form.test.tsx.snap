// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component should render the component 1`] = `
<form
  className="lsux-form"
>
  <ForwardRef(FormFieldComponent)
    id="input"
    label="text"
  >
    <ForwardRef(InputComponent) />
  </ForwardRef(FormFieldComponent)>
</form>
`;
